k8s/pkg/apis/longhorn/v1beta2/objectstore.go

```diff
@@ -0,0 +1,175 @@
+	//  │
+	//  └──►┌───────────┐
+	//      │ running   │
+	//  ┌── └───────────┘
+	//  │
+	//  │ object endpoint is marked for deletion
+	//  │
+	//  └──►┌───────────┐
+	//      │ stopping  │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller waits for dependent resources to disappear before removing
+	//  │ the finalizer, thereby letting the object store be deleted
+	//
+	// +optional
+	// +default="Unknown"
+	State ObjectStoreState `json:"state"`
+
+	// A list of addresses where the S3 store is exposed.
+	//
+	// +optional
+	Endpoints []string `json:"endpoints"`
+}
+
+type ObjectStoreCredentials struct {
+	// An access key. This will be the initial access key for the administrative
+	// account.
+	//
+	// +optional
+	AccessKey string `json:"accessKey"`
+
+	// A secret key. This is the secret key corresponding to the access key above.
+	//
+	// +optional
+	SecretKey string `json:"secretKey"`
+}
+
+type ObjectStoreStorageSpec struct {
+	// The initial size of the volume to provision. The volume will be created
+	// such that it can be expanded if resources are available
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:example:=100Gi
+	// +optional
+	Size resource.Quantity `json:"size"`
+
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default:=3
+	// +optional
+	NumberOfReplicas int `json:"numberOfReplicas"`
+
+	// Replica soft anti affinity of the volume. Set enabled to allow replicas to
+	// be scheduled on the same node
+	//
+	// +optional
+	ReplicaSoftAntiAffinity ReplicaSoftAntiAffinity `json:"replicaSoftAntiAffinity"`
+
+	// Replica zone soft anti affinity of the volume. Set enabled to allow
+	// replicas to be scheduled in the same zone
+	//
+	// +optional
+	ReplicaZoneSoftAntiAffinity ReplicaZoneSoftAntiAffinity `json:"replicaZoneSoftAntiAffinity"`
+
+	// +optional
+	DiskSelector []string `json:"diskSelector"`
+	// +optional
+	NodeSelector []string `json:"nodeSelector"`
+
+	// +optional
+	FromBackup string `json:"fromBackup"`
+
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default:=2880
+	// +optional
+	StaleReplicaTimeout int `json:"staleReplicaTimeout"`
+}
+
+type ObjectEndpointSpec struct {
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// +optional
+	DomainName string `json:"domainName"`
+}
+
+type ObjectStoreSpec struct {
+	// Credentials contain a pair of access and secret keys that are used to seed
+	// the object store.
+	//
+	// +optional
+	Credentials ObjectStoreCredentials `json:"credentials"`
+
+	// The configuration of the longhorn volume providing storage for the object
+	// store.
+	//
+	// +optional
+	Storage ObjectStoreStorageSpec `json:"storage"`
+
+	// A list of endpoint configurations to make the object store available at.
+	//
+	// +optional
+	Endpoints []ObjectEndpointSpec `json:"endpoints"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope="Cluster",shortName={lhos}
```

Comment:
Proposing to change this CRD to have namespace scope. Ref https://github.com/longhorn/longhorn-manager/pull/2135#discussion_r1317906631